File: lib/coffee_script/command_line.rb
require 'optparse'
require 'fileutils'
require 'open3'
require File.expand_path(File.dirname(__FILE__) + '/../coffee-script')

module CoffeeScript

  # The CommandLine handles all of the functionality of the `coffee`
  # utility.
  class CommandLine

    BANNER = <<-EOS
coffee compiles CoffeeScript source files into JavaScript.

Usage:
  coffee path/to/script.coffee
    EOS

    # Seconds to pause between checks for changed source files.
    WATCH_INTERVAL = 0.5

    # Command to execute in Narwhal
    LAUNCHER = "narwhal -e 'require(\"coffee-script\").run(system.args);'"

    # Run the CommandLine off the contents of ARGV.
    def initialize
      @mtimes = {}
      parse_options
      return launch_repl if @options[:interactive]
      return eval_scriptlet if @options[:eval]
      check_sources
      return run_scripts if @options[:run]
      @sources.each {|source| compile_javascript(source) }
      watch_coffee_scripts if @options[:watch]
    end

    # The "--help" usage message.
    def usage
      puts "\n#{@option_parser}\n"
      exit
    end


    private

    # Compiles (or partially compiles) the source CoffeeScript file, returning
    # the desired JS, tokens, or lint results.
    def compile_javascript(source)
      script = File.read(source)
      return tokens(script) if @options[:tokens]
      js = compile(script, source)
      return unless js
      return puts(js) if @options[:print]
      return lint(js) if @options[:lint]
      File.open(path_for(source), 'w+') {|f| f.write(js) }
    end

    # Spins up a watcher thread to keep track of the modification times of the
    # source files, recompiling them whenever they're saved.
    def watch_coffee_scripts
      watch_thread = Thread.start do
        loop do
          @sources.each do |source|
            mtime = File.stat(source).mtime
            @mtimes[source] ||= mtime
            if mtime > @mtimes[source]
              @mtimes[source] = mtime
              compile_javascript(source)
            end
          end
          sleep WATCH_INTERVAL
        end
      end
      Signal.trap("INT") { watch_thread.kill }
      watch_thread.join
    end

    # Ensure that all of the source files exist.
    def check_sources
      usage if @sources.empty?
      missing = @sources.detect {|s| !File.exists?(s) }
      if missing
        STDERR.puts("File not found: '#{missing}'")
        exit(1)
      end
    end

    # Pipe compiled JS through JSLint (requires a working 'jsl' command).
    def lint(js)
      stdin, stdout, stderr = Open3.popen3('jsl -nologo -stdin')
      stdin.write(js)
      stdin.close
      puts stdout.read.tr("\n", '')
      errs = stderr.read.chomp
      puts errs unless errs.empty?
      stdout.close and stderr.close
    end

    # Eval a little piece of CoffeeScript directly from the command line.
    def eval_scriptlet
      script = STDIN.tty? ? @sources.join(' ') : STDIN.read
      return tokens(script) if @options[:tokens]
      js = compile(script)
      return lint(js) if @options[:lint]
      puts js
    end

    # Use Narwhal to run an interactive CoffeeScript session.
    def launch_repl
      exec "#{LAUNCHER}"
    rescue Errno::ENOENT
      puts "Error: Narwhal must be installed to use the interactive REPL."
      exit(1)
    end

    # Use Narwhal to compile and execute CoffeeScripts.
    def run_scripts
      sources = @sources.join(' ')
      exec "#{LAUNCHER} #{sources}"
    rescue Errno::ENOENT
      puts "Error: Narwhal must be installed in order to execute CoffeeScripts."
      exit(1)
    end

    # Print the tokens that the lexer generates from a source script.
    def tokens(script)
      puts Lexer.new.tokenize(script).inspect
    end

    # Compile a single source file to JavaScript.
    def compile(script, source='error')
      begin
        options = {}
        options[:no_wrap] = true if @options[:no_wrap]
        CoffeeScript.compile(script, options)
      rescue CoffeeScript::ParseError, SyntaxError => e
        STDERR.puts "#{source}: #{e.message}"
        exit(1) unless @options[:watch]
        nil
      end
    end

    # Write out JavaScript alongside CoffeeScript unless an output directory
    # is specified.
    def path_for(source)
      filename = File.basename(source, File.extname(source)) + '.js'
      dir      = @options[:output] || File.dirname(source)
      File.join(dir, filename)
    end

    # Install the CoffeeScript TextMate bundle to ~/Library.
    def install_bundle
      bundle_dir = File.expand_path('~/Library/Application Support/TextMate/Bundles/')
      FileUtils.cp_r(File.dirname(__FILE__) + '/CoffeeScript.tmbundle', bundle_dir)
    end

    # Use OptionParser for all the options.
    def parse_options
      @options = {}
      @option_parser = OptionParser.new do |opts|
        opts.on('-i', '--interactive', 'run a CoffeeScript REPL (requires Narwhal)') do |i|
          @options[:interactive] = true
        end
        opts.on('-r', '--run', 'compile and run a script (requires Narwhal)') do |r|
          @options[:run] = true
        end
        opts.on('-o', '--output [DIR]', 'set the directory for compiled JavaScript') do |d|
          @options[:output] = d
          FileUtils.mkdir_p(d) unless File.exists?(d)
        end
        opts.on('-w', '--watch', 'watch scripts for changes, and recompile') do |w|
          @options[:watch] = true
        end
        opts.on('-p', '--print', 'print the compiled JavaScript to stdout') do |d|
          @options[:print] = true
        end
        opts.on('-l', '--lint', 'pipe the compiled JavaScript through JSLint') do |l|
          @options[:lint] = true
        end
        opts.on('-e', '--eval', 'compile a cli scriptlet or read from stdin') do |e|
          @options[:eval] = true
        end
        opts.on('-t', '--tokens', 'print the tokens that the lexer produces') do |t|
          @options[:tokens] = true
        end
        opts.on('-v', '--verbose', 'print at every step of code generation') do |v|
          ENV['VERBOSE'] = 'true'
        end
        opts.on('-n', '--no-wrap', 'raw output, no safety wrapper or vars') do |n|
          @options[:no_wrap] = true
        end
        opts.on_tail('--install-bundle', 'install the CoffeeScript TextMate bundle') do |i|
          install_bundle
          exit
        end
        opts.on_tail('--version', 'display CoffeeScript version') do
          puts "CoffeeScript version #{CoffeeScript::VERSION}"
          exit
        end
        opts.on_tail('-h', '--help', 'display this help message') do
          usage
        end
      end
      @option_parser.banner = BANNER
      begin
        @option_parser.parse!(ARGV)
      rescue OptionParser::InvalidOption => e
        puts e.message
        exit(1)
      end
      @sources = ARGV
    end

  end

end

File: lib/coffee_script/command_line.js
(function(){
  var BANNER, SWITCHES, WATCH_INTERVAL, coffee, compile, compile_scripts, fs, lint, option_parser, options, optparse, parse_options, path, sources, usage, version, write_js;
  fs = require('fs');
  path = require('path');
  coffee = require('coffee-script');
  optparse = require('optparse');
  BANNER = "coffee compiles CoffeeScript source files into JavaScript.\n\nUsage:\n  coffee path/to/script.coffee";
  SWITCHES = [['-i', '--interactive', 'run an interactive CoffeeScript REPL'], ['-r', '--run', 'compile and run a CoffeeScript'], ['-o', '--output [DIR]', 'set the directory for compiled JavaScript'],
    // ['-w', '--watch',         'watch scripts for changes, and recompile']
    ['-p', '--print', 'print the compiled JavaScript to stdout'], ['-l', '--lint', 'pipe the compiled JavaScript through JSLint'], ['-e', '--eval', 'compile a string from the command line'], ['-t', '--tokens', 'print the tokens that the lexer produces'], ['-tr', '--tree', 'print the parse tree that Jison produces'], ['-v', '--version', 'display CoffeeScript version'], ['-h', '--help', 'display this help message']
  ];
  WATCH_INTERVAL = 0.5;
  options = {};
  sources = [];
  option_parser = null;
  // The CommandLine handles all of the functionality of the `coffee` utility.
  exports.run = function run() {
    var flags, separator;
    parse_options();
    if (options.interactive) {
      return require('./repl');
    }
    if (options.eval) {
      return puts(coffee.compile(sources[0]));
    }
    if (!(sources.length)) {
      usage();
    }
    separator = sources.indexOf('--');
    flags = [];
    if (separator >= 0) {
      flags = sources.slice((separator + 1), sources.length);
      sources = sources.slice(0, separator);
    }
    process.ARGV = flags;
    compile_scripts();
    return this;
  };
  // The "--help" usage message.
  usage = function usage() {
    puts('\n' + option_parser.help() + '\n');
    return process.exit(0);
  };
  // The "--version" message.
  version = function version() {
    puts("CoffeeScript version " + coffee.VERSION);
    return process.exit(0);
  };
  // Compile a single source file to JavaScript.
  compile = function compile(script, source) {
    source = source || 'error';
    options = {};
    if (options.no_wrap) {
      options.no_wrap = true;
    }
    if (options.globals) {
      options.globals = true;
    }
    try {
      return CoffeeScript.compile(script, options);
    } catch (error) {
      process.stdio.writeError(source + ': ' + error.toString());
      if (!(options.watch)) {
        process.exit(1);
      }
      return null;
    }
  };
  // Compiles the source CoffeeScript, returning the desired JavaScript, tokens,
  // or JSLint results.
  compile_scripts = function compile_scripts() {
    var opts, source;
    if (!((source = sources.shift()))) {
      return null;
    }
    opts = options;
    return fs.cat(source).addCallback(function(code) {
      var js;
      if (opts.tokens) {
        coffee.print_tokens(coffee.tokenize(code));
      } else if (opts.tree) {
        puts(coffee.tree(code).toString());
      } else {
        js = coffee.compile(code);
        if (opts.run) {
          process.compile(js, source);
        } else if (opts.print) {
          puts(js);
        } else if (opts.lint) {
          lint(js);
        } else {
          write_js(source, coffee.compile(code));
        }
      }
      return compile_scripts();
    });
  };
  // Write out a JavaScript source file with the compiled code.
  write_js = function write_js(source, js) {
    var dir, filename, js_path;
    filename = path.basename(source, path.extname(source)) + '.js';
    dir = options.output || path.dirname(source);
    js_path = path.join(dir, filename);
    return fs.open(js_path, process.O_CREAT | process.O_WRONLY | process.O_TRUNC, parseInt('0755', 8)).addCallback(function(fd) {
      return fs.write(fd, js);
    });
  };
  // Pipe compiled JS through JSLint (requires a working 'jsl' command).
  lint = function lint(js) {
    var jsl;
    jsl = process.createChildProcess('jsl', ['-nologo', '-stdin']);
    jsl.addListener('output', function(result) {
      if (result) {
        return puts(result.replace(/\n/g, ''));
      }
    });
    jsl.addListener('error', function(result) {
      if (result) {
        return puts(result);
      }
    });
    jsl.write(js);
    return jsl.close();
  };
  // Use OptionParser for all the options.
  parse_options = function parse_options() {
    var oparser, opts, paths;
    opts = (options = {});
    oparser = (option_parser = new optparse.OptionParser(SWITCHES));
    oparser.banner = BANNER;
    oparser.add('interactive', function() {
      return opts.interactive = true;
    });
    oparser.add('run', function() {
      return opts.run = true;
    });
    oparser.add('output', function(dir) {
      return opts.output = dir;
    });
    oparser.add('watch', function() {
      return opts.watch = true;
    });
    oparser.add('print', function() {
      return opts.print = true;
    });
    oparser.add('lint', function() {
      return opts.lint = true;
    });
    oparser.add('eval', function() {
      return opts.eval = true;
    });
    oparser.add('tokens', function() {
      return opts.tokens = true;
    });
    oparser.add('tree', function() {
      return opts.tree = true;
    });
    oparser.add('help', (function(__this) {
      var __func = function() {
        return usage();
      };
      return (function() {
        return __func.apply(__this, arguments);
      });
    })(this));
    oparser.add('version', (function(__this) {
      var __func = function() {
        return version();
      };
      return (function() {
        return __func.apply(__this, arguments);
      });
    })(this));
    paths = oparser.parse(process.ARGV);
    return sources = paths.slice(2, paths.length);
  };
})();